lib/bitrange.h

```diff
@@ -51,4 +58,686 @@ class bitranges {
     iter end() const { return iter(bits.end()); }
 };
 
+class JSONGenerator;
+class JSONLoader;
+
+/**
+ * @return the result of dividing @dividend by @divisor, rounded towards
+ * negative infinity. This is different than normal C++ integer division, which
+ * rounds towards zero. For example, `-7 / 8 == 0`, but
+ * `divideFloor(-7, 8) == -1`.
+ */
+inline int divideFloor(int dividend, int divisor) {
+    const int quotient = dividend / divisor;
+    const int remainder = dividend % divisor;
+    if ((remainder != 0) && ((remainder < 0) != (divisor < 0))) return quotient - 1;
+    return quotient;
+}
+
+/**
+ * @return @dividend modulo @divisor as distinct from C++'s
+ * `dividend % divisor`. The difference is that the result of
+ * `dividend % divisor` takes its sign from the dividend, but
+ * `modulo(dividend, divisor)` takes its sign from the divisor.
+ * For example, `-7 % 8 == -7`, but `modulo(-7, 8) == 7`.
+ */
+constexpr int modulo(int dividend, int divisor) {
+    return (dividend % divisor) * ((dividend < 0) != (divisor < 0) ? -1 : 1);
+}
+
+/**
+ * @return the remainder of dividing @dividend by @divisor, rounded towards
+ * negative infinity. This is different than the normal C++ `dividend % divisor` in
+ * two respects:
+ *   - As with `modulo(dividend, divisor)`, it takes its sign from the divisor.
+ *   - For all integers `D1`, `D2`, `D3`,
+ *     `divideFloor(D1, D3) == divideFloor(D2, D3) /\ D1 < D2
+ *          ==> moduloFloor(D1, D3) < moduloFloor(D2, D3)`.
+ *     This is not true for `%`, because due to C++'s round-towards-zero
+ *     behavior the remainders grow in different directions for positive
+ *     dividends than for negative dividends.
+ * To make this concrete, `-7 % 8 == -7`, but `moduloFloor(-7, 8) == 1`.
+ */
+inline int moduloFloor(const int dividend, const int divisor) {
+    const int remainder = modulo(dividend, divisor);
+    if (remainder == 0 || dividend >= 0) return remainder;
+    return divisor - remainder;
+}
+
+/// Units in which a range can be specified.
+enum class RangeUnit : uint8_t {
+    Bit = 0,
+    Byte = 1  /// 8-bit bytes (octets).
+};
+
+/// An ordering for bits or bytes.
+enum class Endian : uint8_t {
+    Network = 0,  /// Most significant bit/byte first.
+    Little = 1,   /// Least significant bit/byte first.
+};
+
+/**
+ * A helper type used to construct a range, specified as the closed interval
+ * between two bit indices. For example, `FromTo(6, 8)` denotes a range
+ * containing three bits: bit 6, bit 7, and bit 8.
+ *
+ * FromTo is only intended to be used as a constructor parameter, to allow
+ * creation of ranges in a more readable manner; use it to construct a
+ * HalfOpenRange or ClosedRange.
+ */
+struct FromTo {
```

Comment:
Can these be put into some namespace? So the usage from the source code would be e.g. `bitrange::FromTo(6, 8)`?